test/e2e/storage/testsuites/snapshottable_stress.go

```diff
@@ -0,0 +1,233 @@
+			e2eskipper.Skipf("Driver %s doesn't implement DynamicPVTestDriver -- skipping", driverInfo.Name)
+		}
+	})
+
+	// This intentionally comes after checking the preconditions because it
+	// registers its own BeforeEach which creates the namespace. Beware that it
+	// also registers an AfterEach which renders f unusable. Any code using
+	// f must run inside an It or Context callback.
+	f := framework.NewDefaultFramework("snapshottable-stress")
+
+	init := func() {
+		cs = f.ClientSet
+		config, driverCleanup := driver.PrepareTest(f)
+		ctx, cancel := context.WithCancel(context.Background())
+
+		stressTest = &snapshottableStressTest{
+			config:        config,
+			driverCleanup: driverCleanup,
+			volumes:       []*VolumeResource{},
+			snapshots:     []*SnapshotResource{},
+			pods:          []*v1.Pod{},
+			testOptions:   *driverInfo.StressTestOptions,
+			ctx:           ctx,
+			cancel:        cancel,
+		}
+	}
+
+	createPodsAndVolumes := func() {
+		for i := 0; i < stressTest.testOptions.NumPods; i++ {
+			framework.Logf("Creating resources for pod %v/%v", i, stressTest.testOptions.NumPods-1)
+
+			volume := CreateVolumeResource(driver, stressTest.config, pattern, t.GetTestSuiteInfo().SupportedSizeRange)
+			stressTest.volumes = append(stressTest.volumes, volume)
+
+			podConfig := e2epod.Config{
+				NS:           f.Namespace.Name,
+				PVCs:         []*v1.PersistentVolumeClaim{volume.Pvc},
+				SeLinuxLabel: e2epv.SELinuxLabel,
+			}
+			pod, err := e2epod.MakeSecPod(&podConfig)
+			framework.ExpectNoError(err)
+			stressTest.pods = append(stressTest.pods, pod)
+
+			if _, err := cs.CoreV1().Pods(pod.Namespace).Create(context.TODO(), pod, metav1.CreateOptions{}); err != nil {
+				stressTest.cancel()
+				framework.Failf("Failed to create pod-%v [%+v]. Error: %v", i, pod, err)
+			}
+
+			if err := e2epod.WaitForPodRunningInNamespace(cs, pod); err != nil {
+				stressTest.cancel()
+				framework.Failf("Failed to wait for pod-%v [%+v] turn into running status. Error: %v", i, pod, err)
+			}
+		}
+	}
+
+	cleanup := func() {
+		framework.Logf("Stopping and waiting for all test routines to finish")
+		stressTest.cancel()
+		stressTest.wg.Wait()
+
+		var errs []error
+		for _, snapshot := range stressTest.snapshots {
+			framework.Logf("Deleting snapshot %v", snapshot.Vs.GetName())
+			errs = append(errs, snapshot.CleanupResource())
+		}
+		for _, pod := range stressTest.pods {
+			framework.Logf("Deleting pod %v", pod.Name)
+			errs = append(errs, e2epod.DeletePodWithWait(cs, pod))
+		}
+		for _, volume := range stressTest.volumes {
+			framework.Logf("Deleting volume %v", volume.Pvc.GetName())
+			errs = append(errs, volume.CleanupResource())
+		}
+		errs = append(errs, tryFunc(stressTest.driverCleanup))
+		framework.ExpectNoError(errors.NewAggregate(errs), "While cleaning up resources")
+	}
+
+	ginkgo.BeforeEach(func() {
+		init()
+		createPodsAndVolumes()
+	})
+
+	f.AddAfterEach("cleanup", func(f *framework.Framework, failed bool) {
```

Comment:
I still find this odd. Most of the other kubernetes tests are using `ginkgo.AfterEach`. Have we checked with test-infra folks if this is expected?

Reply:
I did some more testing using `ginkgo.AfterEach()`, placing before and after we create the `framework`, and could not get it to run.

I observed the same issue with other E2E tests outside of storage. I filed an issue here https://github.com/kubernetes/kubernetes/issues/96177 to figure this out.